plutus-playground/plutus-playground-server/app/Auth.hs

```diff
@@ -0,0 +1,315 @@
+import           Control.Monad.Except        (MonadError)
+import           Control.Monad.IO.Class      (MonadIO, liftIO)
+import           Control.Monad.Logger        (MonadLogger, logDebugN, logErrorN, logInfoN)
+import           Control.Monad.Now           (MonadNow, getCurrentTime, getPOSIXTime)
+import           Control.Monad.Trace         (attempt, runTrace, withTrace)
+import           Control.Monad.Web           (MonadWeb, doRequest, makeManager)
+import           Control.Newtype.Generics    (unpack)
+import           Data.Aeson                  (FromJSON, ToJSON, Value (String), eitherDecode, parseJSON, withObject,
+                                              (.:))
+import           Data.Bifunctor              (first)
+import qualified Data.ByteString.Lazy        as LBS
+import qualified Data.Map                    as Map
+import           Data.Text                   (Text)
+import qualified Data.Text                   as Text
+import           Data.Text.Encoding          (decodeUtf8, encodeUtf8)
+import           Data.Time                   (NominalDiffTime, UTCTime, addUTCTime)
+import           Data.Time.Clock.POSIX       (POSIXTime, utcTimeToPOSIXSeconds)
+import           GHC.Generics                (Generic)
+import           Gist                        (Gist)
+import qualified Gist
+import           Network.HTTP.Client         (managerModifyRequest)
+import           Network.HTTP.Client.Conduit (getUri)
+import           Network.HTTP.Client.TLS     (tlsManagerSettings)
+import           Network.HTTP.Conduit        (Request, newManager, parseRequest, responseBody, responseStatus,
+                                              setQueryString)
+import           Network.HTTP.Simple         (addRequestHeader, getRequestQueryString)
+import           Network.HTTP.Types          (hAccept, statusIsSuccessful)
+import           Servant                     ((:<|>) ((:<|>)), (:>), Get, Header, Headers, JSON, NoContent (NoContent),
+                                              QueryParam, ServantErr, ServerT, StdMethod (GET), ToHttpApiData, Verb,
+                                              addHeader, err401, err500, errBody, throwError)
+import           Servant.API.BrowserHeader   (BrowserHeader)
+import           Servant.Client              (BaseUrl, mkClientEnv, parseBaseUrl, runClientM)
+import           Servant.Extra               ()
+import           Web.Cookie                  (SetCookie, defaultSetCookie, parseCookies, setCookieExpires,
+                                              setCookieHttpOnly, setCookieMaxAge, setCookieName, setCookiePath,
+                                              setCookieSecure, setCookieValue)
+import qualified Web.JWT                     as JWT
+
+data AuthRole
+  = Anonymous
+  | GithubUser
+  deriving (Show, Eq, Generic, FromJSON, ToJSON)
+
+newtype AuthStatus = AuthStatus
+  { _authStatusAuthRole :: AuthRole
+  } deriving (Show, Eq, Generic, FromJSON, ToJSON)
```

Comment:
is this not better of as `newtype AuthStatus = AuthStatus AuthRole` and then using the generic newtype stuff?

Reply:
It is using the generic new type stuff isn't it?

I'm leaving this as a record because I expect to add some more fields before the feature gets merged. (But I've made it a `newtype` while it only has one field to make `hlint` happy. :-)